lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -85,6 +85,24 @@ class SPIRVRegularizeLLVM : public ModulePass {
   /// @spirv.llvm_memset_* and replace it with @llvm.memset.
   void lowerMemset(MemSetInst *MSI);
 
+  /// No SPIR-V counterpart for @llvm.fshl.* intrinsic. It will be lowered
+  /// to a newly generated @spirv.llvm_fshl_* function.
+  ///
+  /// Conceptually, FSHL:
+  /// 1. concatenates the ints, the first one being the more significant;
+  /// 2. performs a left shift-rotate on the resulting doubled-sized int;
+  /// 3. returns the most significant bits of the shift-rotate result,
+  ///    the number of bits being equal to the size of the original integers.
+  /// If FSHL operates on a vector type instead, the same operations are
+  /// performed for each set of corresponding vector elements.
+  ///
+  /// The actual implementation algorithm will be slightly different to speed
+  /// things up.
```

Comment:
I guess this sentence is not actual anymore.

Reply:
I'd argue it is, because I'm performing a left shift + right shift + bitwise addition instead of the algorithm described here.

Reply:
Does it "speed things up"?

Reply:
I'd speculate that a real MSB extraction would require more instructions. I think I should replace it with "simplification".